### PR TITLE
Add cs_buffer. Breaking API change.

### PR DESCRIPTION
**Pull request by numas13**
Remove `cs_malloc()` and `cs_free()`.

`cs_disasm_iter()` is a tiny wrapper around `cs_disasm(count=1)`.

 

**Your checklist for this pull request**
- [x] I've documented or updated the documentation of every API function and struct this PR changes.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)

**Detailed description**

API change unifies disassembly process. Before, there were two separate functions to disassemble one instruction at a time in a loop and many instructions into a dynamic buffer. Commit will introduce user allocatable buffer that can be used in both situations with one function.

Updating the use of `cs_disasm_iter()` is a little more complicated than `cs_disasm()`:

```c
// old api
cs_insn *insn = cs_malloc(handle);
while (cs_disasm_iter(handle, &code, &code_size, &ip, insn)) {
    disassembled_instructions += 1;
}
cs_free(insn);
```

Must be changed to:

```c
// new api
cs_buffer *buffer = cs_buffer_new(1); // create buffer with 1 element
while (cs_disasm_iter(handle, &code, &code_size, &ip, buffer)) {
    cs_insn *insn = &buffer->insn[0]; // get first insn in a buffer
    disassembled_instructions += 1;
}
cs_buffer_free(buffer); // free buffer
```

Updating the use of `cs_disasm()` is straightforward, just use `cs_buffer_new(0)` to create a buffer and pass it to `cs_disasm()`.

**Benchmarks**

```
❯ hyperfine -N ./orig/test_iter_benchmark ./test_iter_benchmark
Benchmark 1: ./orig/test_iter_benchmark
  Time (mean ± σ):     10.632 s ±  0.227 s    [User: 10.622 s, System: 0.004 s]
  Range (min … max):   10.327 s … 11.007 s    10 runs
 
Benchmark 2: ./test_iter_benchmark
  Time (mean ± σ):     10.815 s ±  0.082 s    [User: 10.812 s, System: 0.001 s]
  Range (min … max):   10.695 s … 10.986 s    10 runs
 
Summary
  './orig/test_iter_benchmark' ran
    1.02 ± 0.02 times faster than './test_iter_benchmark'
```

```
❯ ARGS="10 0x028800 0x1517dd libc.so"; hyperfine -N "./orig/test_file_benchmark $ARGS" "./test_file_benchmark $ARGS"
Benchmark 1: ./orig/test_file_benchmark 10 0x028800 0x1517dd libc.so
  Time (mean ± σ):      2.352 s ±  0.035 s    [User: 2.349 s, System: 0.002 s]
  Range (min … max):    2.299 s …  2.388 s    10 runs
 
Benchmark 2: ./test_file_benchmark 10 0x028800 0x1517dd libc.so
  Time (mean ± σ):      2.392 s ±  0.053 s    [User: 2.387 s, System: 0.003 s]
  Range (min … max):    2.329 s …  2.497 s    10 runs
 
Summary
  './orig/test_file_benchmark 10 0x028800 0x1517dd libc.so' ran
    1.02 ± 0.03 times faster than './test_file_benchmark 10 0x028800 0x1517dd libc.so'
```
 
**Test plan**

Update bindings and user software.

**Closing issues**

...

### CODE REVIEW

File: cs.c
Comment (Rot127) [inline cs.c:1120]: Haven't check the changes below yet, but this probably needs a NULL check for `buffer->insn[i] != NULL`

File: cs.c
Comment (Rot127) [inline cs.c:1138]: The `if (...)` can be removed, because the for loop tests it.

File: cs.c
Comment (Rot127) [inline cs.c]:
```suggestion
		size_t diff = required - buffer->capacity;
```

File: cs.c
Comment (Rot127) [inline cs.c:1170]: ~~All these API functions still documentation please. But better wait with it until the other maintainers took a look at it.~~
Haven't seen the header files. All good.

File: include/capstone/capstone.h
Comment (Rot127) [inline include/capstone/capstone.h:745]: Please document the detail that it at least reserves 16 elements.

**Review by Rot127 — CHANGES_REQUESTED** [review]
Looks awesome! It really cleans up the API a lot. Especially internally.
Please rebase this on onto the ASAN PR from me or wait until it is merged: https://github.com/capstone-engine/capstone/pull/2313
Just so we get extra test coverage.

@aquynh @kabeor Please take a look. I am  very much in favor of this.

File: cs.c
Comment (Rot127) [inline cs.c]: Couldn't you also reserve the buffer once, on based on some good guess how many instructions will be disassembled? So most often `code_size / 4`, sometimes `code_size / 2`.

Comment (numas13) [reply cs.c]: Good idea!

File: cs.c
Comment (numas13) [reply cs.c:1138]: `buffer->insn[i].detail` can be allocated after previous run of `cs_disas()` and needs to be freed if capacity reduced.

File: cs.c
Comment (numas13) [reply cs.c:1120]: I'm not sure. `buffer->insn` is not an array of pointers to `cs_insn`.

**Comment by numas13** [comment]
> Please rebase this on onto the ASAN PR from me or wait until it is merged: #2313 Just so we get extra test coverage.

There is one problem in `test_arm_regression`. Is it expected?

**Comment by numas13** [comment]
I've updated python bindings and review is appreciated.

File: cs.c
Comment (numas13) [reply cs.c:1138]: I've refactored `cs_buffer_reserve_exact()` and fixed memory leak if `cs_mem_realloc()` fails.

Comment (Rot127) [reply cs.c:1138]: Ok, let's wait for @aquynh and @kabeor commenting.

File: cs.c
Comment (Rot127) [reply cs.c:1120]: You are right. Was stuck in the old implementation.

**Comment by aquynh** [comment]
please avoid breaking API

**Comment by numas13** [comment]
I'm closing PRs. We'll think about what can be done.

**Pull request closed without merging**
